src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp

```diff
@@ -1307,24 +1320,181 @@ void WholeBodyDynamicsDevice::readContactPoints()
 
-    for(size_t subModel = 0; subModel < nrOfSubModels; subModel++)
+    // read skin
+    iCub::skinDynLib::skinContactList *scl =this->portContactsInput.read(false); //scl=null could also mean no new message
+    if(scl && useSkinContact)
     {
-        bool ok = measuredContactLocations.addNewContactInFrame(estimator.model(),
-                                                                subModelIndex2DefaultContact[subModel],
-                                                               iDynTree::UnknownWrenchContact(iDynTree::FULL_WRENCH,iDynTree::Position::Zero()));
-        if( !ok )
+
+        //< \todo TODO check for envelope?
+        last_reading_skin_contact_list_Stamp = yarp::os::Time::now();
+        if(scl->empty())   // if no skin contacts => leave the old contacts but reset pressure and contact list
+        {
+
+
+            if (contactsReadFromSkin.empty())
+            {
+                yInfo() << "wholeBodyDynamics: attempting to use previous contacts but previous contacts is empty using default contacts instead";
+                for(size_t subModel = 0; subModel < nrOfSubModels; subModel++)
+                {
+                    bool ok = measuredContactLocations.addNewContactInFrame(estimator.model(),
+                                                                            subModelIndex2DefaultContact[subModel], //frameIndex in iDynTree
+                                                                            iDynTree::UnknownWrenchContact(iDynTree::FULL_WRENCH,iDynTree::Position::Zero()));
+                    if( !ok )
+                    {
+                        yWarning() << "wholeBodyDynamics: Failing in adding default contact for submodel " << subModel;
+                    }
+
+
+
+
+                }
+                return;
+            }
+
+            //< \todo TODO this (using the last contacts if no contacts are detected) should be at subtree level, not at global level??
+            // ask what is intended with this TODO
+            else
+            {
+                for(iCub::skinDynLib::skinContactList::iterator it=contactsReadFromSkin.begin(); it!=contactsReadFromSkin.end(); it++)
+                {
+                    it->setPressure(0.0);
+                    it->setActiveTaxels(0);
+                    yDebug() << "wholeBodyDynamics: skincontactlist empty, setting pressure and active taxels to 0";
+                    numberOfContacts++;
+                }
+                yInfo() << "wholeBodyDynamics: numberOfContacts in contactsReadFromSkin from previous contacts = "<<numberOfContacts;
+
+            }
+
+
+        }
+        else
+        {
+            contactsReadFromSkin.clear();
+            yDebug() << "wholeBodyDynamics: clear contactsReadFromSkin ";
+            int min_taxel=5;//might become a configuration variable
+            //consider only contacts with a minimun of activated taxels, is this neccesary or correct?
```

Comment:
even if required i would argue that it should be deal before arriving to this point, probably by skin manager.